lms/djangoapps/bookmarks_api/views.py

```diff
@@ -0,0 +1,210 @@
+from rest_framework.authentication import OAuth2Authentication, SessionAuthentication
+from rest_framework.generics import ListCreateAPIView
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from bookmarks_api import serializers
+from openedx.core.lib.api.serializers import PaginationSerializer
+
+from xmodule.modulestore.exceptions import ItemNotFoundError
+from xmodule.modulestore.django import modulestore
+
+from .models import Bookmark
+from .api import get_bookmark
+
+
+log = logging.getLogger(__name__)
+
+
+class BookmarksView(ListCreateAPIView):
+    """
+    List all bookmarks or create.
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (permissions.IsAuthenticated,)
+
+    paginate_by = 1000
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = serializers.BookmarkSerializer
+
+    def get_queryset(self):
+        course_id = self.request.QUERY_PARAMS.get('course_id', None)
+        fields = self.request.QUERY_PARAMS.get('fields', None)
+
+        if not course_id:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            log.error("Invalid course id '{course_id}'")
+            return list()
+
+        results_queryset = Bookmark.objects.filter(course_key=course_key, user__id=self.request.user.id).values("course_key", "usage_key", "display_name", "created").order_by('-created')
+        # serializer = serializers.BookmarkSerializer(results, many=True, fields_to_remove=['path'])
+
+        return results_queryset
+
+    def post(self, request):
+        """
+        Create a new bookmark.
+
+        Returns 400 request if bad payload is sent or it was empty object.
+        """
+        if not request.DATA:
+            error_message = _("No data provided for bookmark")
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        usage_id = request.DATA.get('usage_id', None)
+        if not usage_id:
+            error_message = _('No usage id provided for bookmark')
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            usage_key = UsageKey.from_string(usage_id)
+
+            # usage_key's course_key may have an empty run property
+            usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
```

Comment:
We don't need to add check here because that's already in `fill_in_run` method. you can see here https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/mongo/base.py#L620